### ISSUE TITLE
Cursor pagination bug with relationships when ordering by non-primary key column

### ISSUE THREAD

**Issue by stefansundin** [mikro-orm/mikro-orm#5155]
### Describe the bug

When performing cursor pagination with relationships, the cursors returned are incomplete if the ordering column is not the primary key.

The first lookup below returns a cursor that decodes to `[{"user":{}}` when it should be `[{"user":{"name":"Bar"}}`. Performing a lookup on `User` in the middle seems to fix the problem so the second lookup returns a good cursor. It also fixes the cursors of the first query, presumably because some metadata loaded because of the `User` lookup.

The tests below works if you change `name` to `id`, i.e. `{ orderBy: { user: { name: QueryOrder.ASC } }` to `{ orderBy: { user: { id: QueryOrder.ASC } }`.

Anyway, a repro is better than my attempt to describe the bug, so here's a repro:

```typescript
test('basic CRUD example', async () => {
  const org = orm.em.create(Org, { name: 'The Org' });
  const user1 = orm.em.create(User, { name: 'Foo', email: 'foo' });
  const user2 = orm.em.create(User, { name: 'Bar', email: 'bar' });
  const user3 = orm.em.create(User, { name: 'Baz', email: 'baz' });
  orm.em.create(OrgMembership, { user: user1, org });
  orm.em.create(OrgMembership, { user: user2, org });
  orm.em.create(OrgMembership, { user: user3, org });
  await orm.em.flush();
  orm.em.clear();

  // This cursor, W3sidXNlciI6e319XQ, decodes to [{"user":{}}, which is wrong
  const badCursor = await orm.em.findByCursor(OrgMembership, {}, { orderBy: { user: { name: QueryOrder.ASC } }, first: 1 });
  console.log({ badCursor: badCursor.endCursor });
  const badCursorEndCursorCopy = badCursor.endCursor; // making a copy here because the next query will change the endCursor

  // Uncomment this when the bug is fixed:
  // expect(badCursor.endCursor).toBe("W3sidXNlciI6eyJuYW1lIjoiQmFyIn19XQ");

  // Performing this query will make it so that goodCursor receives a valid cursor
  // Comment this out and goodCursor will receive the same bad cursor as badCursor
  // This query also changes the output of badCursor.endCursor, probably because some metadata is loaded behind the scenes?
  const dummyUserLookup = await orm.em.findByCursor(User, {}, { orderBy: { name: QueryOrder.ASC }, first: 1 });

  // This cursor, W3sidXNlciI6eyJuYW1lIjoiQmFyIn19XQ, decodes to [{"user":{"name":"Bar"}}, which is expected
  const goodCursor = await orm.em.findByCursor(OrgMembership, {}, { orderBy: { user: { name: QueryOrder.ASC } }, first: 1 });
  console.log({ goodCursor: goodCursor.endCursor });

  // These two cursors should both be the same, and not change over time
  expect(goodCursor.endCursor).toBe("W3sidXNlciI6eyJuYW1lIjoiQmFyIn19XQ");
  expect(badCursor.endCursor).toBe(goodCursor.endCursor);
  expect(badCursor.endCursor).toBe(badCursorEndCursorCopy);
});
```

### Reproduction

https://github.com/stefansundin/mikro-orm-reproduction/tree/cursor-relationship-bug

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.0.4

### Node.js version

v18.19.0

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
The problem here is that you don't populate the relation, the populate hint is not inferred from order by clause automatically. It works fine if yo do this:

```ts
const badCursor = await orm.em.findByCursor(OrgMembership, {}, { 
  populate: ['user'],
  orderBy: { user: { name: QueryOrder.ASC } }, 
  first: 1,
});
```

Without populating the entity, there is no value for `user.name`, so it resolves to `undefined` and the key is removed when serializing to JSON.

It feels like a wontfix to me, the order by clause shouldn't force population automatically. Maybe we could throw in case the value relation is not populated.

**Comment by stefansundin** [comment]
That makes a lot of sense. I feel dumb not realizing that myself. Thank you. Perhaps it would be good to freeze the cursor string so that they can't be modified afterwards, but that's another discussion.

I have another question. I know that using cursor pagination with non-unique columns like names are not generally a good idea, but sometimes you still want to provide the user with the option to sort the data on arbitrary columns.

For example, this does not work well (two users with the name `Bar`):

```typescript
test('basic CRUD example', async () => {
  const org = orm.em.create(Org, { name: 'The Org' });
  const user1 = orm.em.create(User, { name: 'Abc', email: 'foo' });
  const user2 = orm.em.create(User, { name: 'Bar', email: 'bar' });
  const user3 = orm.em.create(User, { name: 'Bar', email: 'bar2' });
  const user4 = orm.em.create(User, { name: 'Baz', email: 'baz' });
  orm.em.create(OrgMembership, { user: user1, org });
  orm.em.create(OrgMembership, { user: user2, org });
  orm.em.create(OrgMembership, { user: user3, org });
  orm.em.create(OrgMembership, { user: user4, org });
  await orm.em.flush();
  orm.em.clear();

  const cursor1 = await orm.em.findByCursor(OrgMembership, {}, {
    populate: ['user'],
    orderBy: { user: { name: QueryOrder.ASC } },
    first: 1,
  });
  console.log({ cursor1: cursor1.items[0].user.$ });
  expect(cursor1.items[0].user.$.email).toBe('foo');

  const cursor2 = await orm.em.findByCursor(OrgMembership, {}, {
    populate: ['user'],
    orderBy: { user: { name: QueryOrder.ASC } },
    first: 1,
    after: cursor1,
  });
  console.log({ cursor2: cursor2.items[0].user.$ });
  expect(cursor2.items[0].user.$.email).toBe('bar');

  const cursor3 = await orm.em.findByCursor(OrgMembership, {}, {
    populate: ['user'],
    orderBy: { user: { name: QueryOrder.ASC } },
    first: 1,
    after: cursor2,
  });
  console.log({ cursor3: cursor3.items[0].user.$ });
  expect(cursor3.items[0].user.$.email).toBe('bar2');
});
```

The user with `email: 'bar2'` is skipped because the `where` clause is `name > 'Bar'`.

But if the `where` clause could be `(name > 'Bar' or (name = 'Bar' and id > 2))` then it would work. Perhaps `orderBy: { user: { name: QueryOrder.ASC, id: QueryOrder.ASC } }` could be a special syntax with cursor pagination, i.e. when the primary key is sorted last then it magically transforms the `where` clause into this type.

The reason for the extra parenthesis is to account for when the user added their own conditions, in this example it could be `{ org: { id: 1 } }`.

i.e. it would be cool if this would work:

```typescript
test('basic CRUD example', async () => {
  const org = orm.em.create(Org, { name: 'The Org' });
  const user1 = orm.em.create(User, { name: 'Abc', email: 'foo' });
  const user2 = orm.em.create(User, { name: 'Bar', email: 'bar' });
  const user3 = orm.em.create(User, { name: 'Bar', email: 'bar2' });
  const user4 = orm.em.create(User, { name: 'Baz', email: 'baz' });
  orm.em.create(OrgMembership, { user: user1, org });
  orm.em.create(OrgMembership, { user: user2, org });
  orm.em.create(OrgMembership, { user: user3, org });
  orm.em.create(OrgMembership, { user: user4, org });
  await orm.em.flush();
  orm.em.clear();

  const cursor1 = await orm.em.findByCursor(OrgMembership, {}, {
    populate: ['user'],
    orderBy: { user: { name: QueryOrder.ASC, id: QueryOrder.ASC } },
    first: 1,
  });
  console.log({ cursor1: cursor1.items[0].user.$ });
  expect(cursor1.items[0].user.$.email).toBe('foo');

  const cursor2 = await orm.em.findByCursor(OrgMembership, {}, {
    populate: ['user'],
    orderBy: { user: { name: QueryOrder.ASC, id: QueryOrder.ASC } },
    first: 1,
    after: cursor1,
  });
  console.log({ cursor2: cursor2.items[0].user.$ });
  expect(cursor2.items[0].user.$.email).toBe('bar');

  // This where clause is: where `u2`.`name` > 'Bar' and `u2`.`id` > 2
  // Is there a way to make it the following instead?
  // where (`u2`.`name` > 'Bar' or (`u2`.`name` = 'Bar' and `u2`.`id` > 2))
  const cursor3 = await orm.em.findByCursor(OrgMembership, {}, {
    populate: ['user'],
    orderBy: { user: { name: QueryOrder.ASC, id: QueryOrder.ASC } },
    first: 1,
    after: cursor2,
  });
  console.log({ cursor3: cursor3.items[0].user.$ });
  expect(cursor3.items[0].user.$.email).toBe('bar2');
});
```

Let me know what you think about this idea. I think it might be hard or impossible to implement this on top of the existing cursor pagination without modifying mikro-orm itself.

**Comment by B4nan** [comment]
> But if the where clause could be (name > 'Bar' or (name = 'Bar' and id > 2)) then it would work. Perhaps orderBy: { user: { name: QueryOrder.ASC, id: QueryOrder.ASC } } could be a special syntax with cursor pagination, i.e. when the primary key is sorted last then it magically transforms the where clause into this type.

This needs to be explicit, if you want to order by the PK, put it into the `orderBy` clause. I don't see why this should work automatically.

**Comment by stefansundin** [comment]
Makes sense to configure it explicitly.

**Comment by B4nan** [comment]
> Perhaps it would be good to freeze the cursor string so that they can't be modified afterwards, but that's another discussion.

I checked this in detail, and its also the same problem with the populate. The value for `startCursor` and `endCursor` is computed (they are getters) from the loaded items, and since you did not have the user-populated, the computed value was wrong. Once you did the additional `find(User)` query, it became populated and the cursor values changed (to the correct value, since now the user relation was correctly loaded).

**Comment by B4nan** [comment]
> But if the where clause could be (name > 'Bar' or (name = 'Bar' and id > 2)) then it would work. Perhaps orderBy: { user: { name: QueryOrder.ASC, id: QueryOrder.ASC } } could be a special syntax with cursor pagination, i.e. when the primary key is sorted last then it magically transforms the where clause into this type.

This already works fine if you provide an array for the orderBy:

```ts
const cursor3 = await orm.em.findByCursor(OrgMembership, {}, {
  populate: ['user'],
  orderBy: [
    { user: { name: QueryOrder.ASC } },
    { id: 'asc' },
  ],
  first: 1,
  after: cursor2,
});
```

Produces the following query:

```sql
select `o0`.*, `u1`.`id` as `u1__id`, `u1`.`name` as `u1__name`, `u1`.`email` as `u1__email`
from `org_membership` as `o0` left join `user` as `u1` on `o0`.`user_id` = `u1`.`id`
left join `user` as `u2` on `o0`.`user_id` = `u2`.`id`
where `u2`.`name` >= 'Bar' and (`u2`.`name` > 'Bar' or `u2`.`id` > 2)
order by `u2`.`name` asc, `o0`.`user_id` asc
limit 2
```

I found one small issue, doing this fails:

```ts
const cursor3 = await orm.em.findByCursor(OrgMembership, {}, {
  populate: ['user'],
  orderBy: [
    { user: { name: QueryOrder.ASC } },
    { user: 'asc' }, // but `{ user: { id: 'asc' } }` works
  ],
  first: 1,
  after: cursor2,
});
```

Will try to fix that and add the validation for the missing populate hint.
